lib/impl/refactorRules.js

```diff
@@ -0,0 +1,384 @@
+var async = require('async');
+var getSectionMap = require('../utils/get-section-map');
+var _ = require('underscore');
+var models = require('../common/models.js')();
+
+/*
+  Checking rule targets are still valid -- if not, rule is deleted
+*/
+module.exports = function refactorRules(connections, form, ruleDeletionFlags, cb) {
+  //Compiling a list of field and page ids to check for
+  var formPages = [];
+  var fieldRulesToDelete = [];
+  var pageRulesToDelete = [];
+
+  /**
+   * Object of invalid fields that is built up by scanning the rules.
+   * @type {{
+   *   fieldId: true
+   * }}
+   */
+  var invalidFields = {
+
+  };
+
+  //If a rule contains a field that is flagged for rule deletion, it is added to this list.
+  // Invalid field rules with fields within repeating section are also added to this list.
+  var rulesFlaggedForDeletion = {};
+
+  var conn = connections.mongooseConnection;
+  var formModel = models.get(conn, models.MODELNAMES.FORM);
+  var fieldModel = models.get(conn, models.MODELNAMES.FIELD);
+  var fieldRuleModel = models.get(conn, models.MODELNAMES.FIELD_RULE);
+  var pageRuleModel = models.get(conn, models.MODELNAMES.PAGE_RULE);
+
+  formModel.findById(form._id.toString()).populate("pages", "-__v")
+    .populate("pageRules", "-__v")
+    .populate("fieldRules", "-__v")
+    .populate("dataSources.formDataSources", "_id name")//Populating Data Sources
+    .populate("dataTargets", "_id name")//Populating Data Targets
+    .select("-__v").exec(function(err, formResult) {
+      if (err) {
+        return cb(err);
+      }
+
+      //Now have a form with all pages, pageRules and fieldRules populated
+      formModel.populate(formResult, {
+        "path": "pages.fields",
+        "model": fieldModel,
+        "select": "-__v"
+      }, function(err, updatedFormResult) {
+        if (err) {
+          return cb(err);
+        }
+
+        form = updatedFormResult;
+
+        formPages = updatedFormResult.pages;
+
+        async.parallel([
+          function(fieldRuleCb) {
+            async.each(formResult.fieldRules, async.apply(scanForInvalidFields, models.MODELNAMES.FIELD_RULE), fieldRuleCb);
+          },
+          function(pageRuleCb) {
+            async.each(formResult.pageRules, async.apply(scanForInvalidFields, models.MODELNAMES.PAGE_RULE), pageRuleCb);
+          }
+        ], function(err) {
+          if (err) {
+            return cb(err);
+          }
+
+          checkRulesWithRepeatingSections();
+
+          //Update the form object
+          async.series([
+            updateFormRules,
+            updateOrDeleteRules
+          ], cb);
+        });
+      });
+    });
+
+
+  function scanForInvalidFields(type, rule, callback) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Function has a complexity of 22 which is greater than 20 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AFunctionComplexity)